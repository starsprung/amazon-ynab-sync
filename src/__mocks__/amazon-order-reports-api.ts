export const mocks = {
  getItems: jest.fn(),
  getRefunds: jest.fn(),
  stop: jest.fn(),
};

export const AmazonOrderReportsApi = jest.fn(() => mocks);
