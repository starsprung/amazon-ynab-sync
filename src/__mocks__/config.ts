import { normalize } from 'path';
import { Config } from '../config';

const mocks = {
  amazonOtpCode: jest.fn(() => '54321'),
  amazonPassword: jest.fn(() => 'pass123456'),
  amazonUsername: jest.fn(() => 'user@example.com'),
  cacheDir: jest.fn(() => normalize('/path/to/cache/')),
  cleared: jest.fn(() => true),
  logLevel: jest.fn(() => 'none'),
  payee: jest.fn(() => undefined),
  ynabAccessToken: jest.fn(() => 'f82918ba-4aa7-4805-b9be-fe5e87eaacf3'),
  ynabAccountName: jest.fn(() => 'Amazon.com'),
  ynabBudgetName: jest.fn(() => 'Budget'),
};

const config = Object.defineProperties(
  {},
  Object.fromEntries(Object.entries(mocks).map(([k, v]) => [k, { get: v, configurable: true }])),
);

export const getConfig = jest.fn(() => config as Config);
