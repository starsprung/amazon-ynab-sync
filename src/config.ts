import { version } from '../package.json';
import envPaths from 'env-paths';
import { readFileSync } from 'fs';
import { join } from 'path';
import yargs from 'yargs';

const CONFIG_FILE = 'config.json';

export interface Config {
  amazonUsername?: string;
  amazonPassword?: string;
  amazonOtpCode?: string;
  amazonOtpSecret?: string;
  cacheDir: string;
  configDir: string;
  debugMode: boolean;
  logLevel: string;
  startDate: string;
  ynabBudgetName: string;
  ynabAccountName: string;
  ynabAccessToken: string;
}

let config: Config;

export const getConfig = (): Config => {
  if (config) {
    return config;
  }

  const { cache: defaultCacheDir, config: defaultConfigDir } = envPaths('amazon-ynab-sync', {
    suffix: '',
  });

  return (config = {
    ...((yargs(process.argv.slice(2))
      .usage('Usage: $0 [options]')
      .env()
      .version(version)
      .option('amazon-otp-code', {
        describe: 'Amazon OTP/2SV code',
        type: 'string',
      })
      .option('amazon-otp-secret', {
        describe: 'Amazon OTP/2SV secret',
        type: 'string',
      })
      .option('amazon-password', {
        describe: 'Amazon account password',
        type: 'string',
      })
      .option('amazon-username', {
        describe: 'Amazon account username',
        type: 'string',
      })
      .option('cache-dir', {
        default: defaultCacheDir,
        describe: 'Location of cache directory',
        type: 'string',
      })
      .option('config-dir', {
        config: true,
        configParser: (path: string): Partial<Config> => {
          try {
            const content = readFileSync(join(path, CONFIG_FILE), { encoding: 'utf-8' });
            return JSON.parse(content);
          } catch {
            return {};
          }
        },
        default: defaultConfigDir,
        describe: 'Location of config directory',
        type: 'string',
      })
      .option('debug-mode', {
        default: false,
        describe: 'Run internal browser in visible mode and run in slo-mo mode',
        type: 'boolean',
      })
      .option('log-level', {
        choices: ['debug', 'info', 'error', 'none', 'silly'],
        default: 'info',
        describe: 'Level of logs to output',
        type: 'string',
      })
      .options('start-date', {
        describe: 'Sync transactions that occur after this date',
        default: (() => {
          const d = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
          return [
            d.getFullYear(),
            `${d.getMonth() + 1}`.padStart(2, '0'),
            `${d.getDate()}`.padStart(2, '0'),
          ].join('-');
        })(),
        type: 'string',
      })
      .option('ynab-access-token', {
        demandOption: true,
        describe: 'YNAB access token for accessing your account',
        type: 'string',
      })
      .option('ynab-account-name', {
        demandOption: true,
        describe: 'Account to use for storing Amazon transactions',
        type: 'string',
      })
      .option('ynab-budget-name', {
        demandOption: true,
        describe: 'Budget to use for storing Amazon transactions',
        type: 'string',
      }).argv as unknown) as Config),
  });
};
