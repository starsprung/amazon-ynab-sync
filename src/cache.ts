import { readFile, writeFile } from 'fs/promises';
import hasha from 'hasha';
import makeDir from 'make-dir';
import { join } from 'path';
import { getConfig } from './config';
import logger from './logger';

// eslint-disable-next-line
type Serializable = string | number | boolean | object | null;

const config = getConfig();

export const makeCacheDir = async (): Promise<void> => {
  logger.debug(`Cache dir is: ${config.cacheDir}`);
  await makeDir(config.cacheDir);
};

const cacheFilePath = (keys: string | Array<string>) => {
  const path = join(config.cacheDir, `${hasha(keys, { algorithm: 'sha256' })}.cache`);
  logger.debug(`Path for ${keys} is ${path}`);
  return path;
};

export const readCache = async (
  keys: string | Array<string>,
  maxAge: number = Number.MAX_SAFE_INTEGER,
): Promise<Serializable> => {
  logger.debug(`Reading ${keys} from cache`);

  try {
    const content = await readFile(cacheFilePath(keys), {
      encoding: 'utf-8',
    });
    if (content) {
      logger.debug(`Found cached value for ${keys}`);
      logger.silly(content);

      const { value, time } = JSON.parse(content);
      const age = Date.now() - parseInt(time);
      logger.debug(`Cached value for ${keys} age is ${age} ms`);
      if (age <= maxAge) {
        return value;
      }

      logger.debug(`Cached value for ${keys} is too old, ignoring`);
    }
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err;
    }
  }

  logger.debug(`No cache value found for ${keys}`);

  return null;
};

export const writeCache = async (
  keys: string | Array<string>,
  value: Serializable,
): Promise<void> => {
  logger.debug(`Writing ${keys} to cache`);
  logger.silly(value);

  await writeFile(cacheFilePath(keys), JSON.stringify({ value, time: Date.now() }, null, 2), {
    encoding: 'utf-8',
  });
};
