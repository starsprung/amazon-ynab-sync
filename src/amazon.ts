import { AmazonOrderReportsApi, Cookie, LogLevel } from 'amazon-order-reports-api';
import hasha from 'hasha';
import inquirer from 'inquirer';
import { SaveTransaction } from 'ynab';
import { readCache, writeCache } from './cache';
import { getConfig } from './config';

const config = getConfig();

export type AmazonTransaction = Omit<SaveTransaction, 'account_id'>;
type IdRecord = Record<string, number>;

const prompt = async (message: string, type: 'input' | 'password' = 'input') =>
  (await inquirer.prompt([{ name: 'p', message, type }]))?.p;

const getUsername = async (): Promise<string> =>
  config.amazonUsername ?? (await prompt('Amazon Username:'));
const getPassword = async (): Promise<string> =>
  config.amazonPassword ?? (await prompt('Amazon Password:', 'password'));
const getOtpCode = async (): Promise<string> =>
  config.amazonOtpCode ?? (await prompt('Amazon OTP Code:'));

const generateImportId = (idInputs: Array<string | number | Date>, seenIds: IdRecord): string => {
  const baseHash = hasha(
    idInputs.map((v) => `${v instanceof Date ? v.toISOString() : v}`),
    { algorithm: 'sha256' },
  );

  const occurence = seenIds[baseHash] ?? 0;
  seenIds[baseHash] = occurence + 1;

  return hasha([baseHash, `${occurence}`], { algorithm: 'sha256' }).substr(0, 36);
};

const createTransation = (
  seenIds: IdRecord,
  type: 'item' | 'refund',
  orderId: string,
  date: Date,
  asinIsbn: string,
  title: string,
  seller: string,
  quantity: number,
  amount: number,
): AmazonTransaction => ({
  amount: Math.round(amount * 1000),
  date: `${date.toISOString().split('T')[0]}`,
  import_id: generateImportId(
    [type, orderId, date, asinIsbn, title, seller, quantity, amount],
    seenIds,
  ),
  memo: title ? `${quantity > 1 ? `${quantity} x ` : ''}${title}`.substr(0, 200) : '',
  payee_name: seller,
});

export const getAmazonTransactions = async function* (): AsyncGenerator<AmazonTransaction> {
  const seenIds: IdRecord = {};

  const api = new AmazonOrderReportsApi({
    username: getUsername,
    password: getPassword,
    otpSecret: config.amazonOtpSecret,
    otpFn: getOtpCode,
    logLevel: config.logLevel as LogLevel,
    cookies: ((await readCache('cookies')) as Array<Cookie>) ?? [],
    saveCookiesFn: (cookies: Array<Cookie>) => writeCache('cookies', cookies),
    puppeteerOpts: config.debugMode ? { headless: false, slowMo: 100 } : {},
  });

  try {
    for await (const item of api.getItems({
      startDate: new Date(config.startDate),
      endDate: new Date(),
    })) {
      if (item.itemTotal) {
        yield createTransation(
          seenIds,
          'item',
          item.orderId,
          item.orderDate,
          item.asinIsbn,
          item.title,
          item.seller,
          item.quantity,
          -item.itemTotal,
        );
      }
    }

    for await (const refund of api.getRefunds({
      startDate: new Date(config.startDate),
      endDate: new Date(),
    })) {
      if (refund.refundAmount) {
        yield createTransation(
          seenIds,
          'refund',
          refund.orderId,
          refund.orderDate,
          refund.asinIsbn,
          refund.title,
          refund.seller,
          refund.quantity,
          refund.refundAmount,
        );
      }
    }
  } finally {
    if (config.debugMode) {
      await new Promise((resolve) => setTimeout(resolve, 10000));
    }

    await api.stop();
  }
};
