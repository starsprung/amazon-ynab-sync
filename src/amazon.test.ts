import { OrderItem, Refund, Shipment } from 'amazon-order-reports-api';
import inquirer from 'inquirer';
import mockdate from 'mockdate';
import { mocked } from 'ts-jest/utils';
import { AmazonTransaction, getAmazonTransactions } from './amazon';
import { readCache, writeCache } from './cache';
import { getConfig } from './config';
import {
  AmazonOrderReportsApi,
  mocks as amazonOrderReportsApiMocks,
} from './__mocks__/amazon-order-reports-api';

jest.mock('inquirer');
jest.mock('./cache');
jest.mock('./config');

describe('amazon', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockdate.set('2021-01-31');
  });

  afterEach(() => {
    mockdate.reset();
  });

  describe('getAmazonTransactions', () => {
    beforeEach(() => {
      mocked(amazonOrderReportsApiMocks).getItems.mockImplementationOnce(async function* () {
        yield* [
          {
            asinIsbn: 'B06Y5XG33Q',
            buyerName: 'Test Man',
            carrierNameTrackingNumber: 'USPS(9300120111405739275629)',
            category: 'HEALTH_PERSONAL_CARE',
            condition: 'new',
            currency: 'USD',
            itemSubtotal: 21.75,
            itemSubtotalTax: 0,
            itemTotal: 21.75,
            listPricePerUnit: 0,
            orderDate: new Date('2020-01-03T08:00:00.000Z'),
            orderId: '123-7654321-1234567',
            orderStatus: 'Shipped',
            orderingCustomerEmail: 'email@example.com',
            paymentInstrumentType: 'Visa - 1234',
            purchasePricePerUnit: 21.75,
            quantity: 1,
            seller: 'This Is A Seller',
            shipmentDate: new Date('2020-01-06T08:00:00.000Z'),
            shippingAddressCity: 'Washington',
            shippingAddressName: 'Test Man',
            shippingAddressState: 'DC',
            shippingAddressStreet1: '1600 Pennsylvania Avenue NW',
            shippingAddressZip: '20500',
            title: 'Some Product',
            unspscCode: '38462846',
            website: 'Amazon.com',
          } as OrderItem,
        ];
      });

      mocked(amazonOrderReportsApiMocks.getRefunds).mockImplementationOnce(async function* () {
        yield* [
          {
            asinIsbn: 'B07USNDYEK',
            buyerName: 'Test Man',
            category: 'COAT',
            orderDate: new Date('2020-01-14T08:00:00.000Z'),
            orderId: '113-5757575-3939393',
            quantity: 2,
            refundAmount: 75.98,
            refundCondition: 'Completed',
            refundDate: new Date('2020-01-25T08:00:00.000Z'),
            refundReason: 'Customer Return',
            refundTaxAmount: 0,
            seller: 'A Seller',
            title: 'Some kind of coat',
            website: 'Amazon.com',
          } as Refund,
          {
            asinIsbn: 'B07USNDYEK',
            buyerName: 'Test Man',
            category: 'COAT',
            orderDate: new Date('2020-01-14T08:00:00.000Z'),
            orderId: '113-5757575-3939393',
            quantity: 2,
            refundAmount: 75.98,
            refundCondition: 'Completed',
            refundDate: new Date('2020-01-25T08:00:00.000Z'),
            refundReason: 'Customer Return',
            refundTaxAmount: 0,
            seller: 'A Seller',
            title: 'Some kind of coat',
            website: 'Amazon.com',
          } as Refund,
        ];
      });

      mocked(amazonOrderReportsApiMocks).getShipments.mockImplementationOnce(async function* () {
        yield* [
          {
            buyerName: 'Test Man',
            carrierNameTrackingNumber: 'FEDEX(536485026870)',
            orderDate: new Date('2020-02-22T08:00:00.000Z'),
            orderId: '112-1234569-3333333',
            orderStatus: 'Shipped',
            orderingCustomerEmail: 'test@example.com',
            paymentInstrumentType: 'American Express - 1236',
            shipmentDate: new Date('2020-02-23T08:00:00.000Z'),
            shippingAddressCity: 'Washington',
            shippingAddressName: 'Test Man',
            shippingAddressState: 'DC',
            shippingAddressStreet1: '1600 Pennsylvania Avenue NW',
            shippingAddressZip: '20500',
            shippingCharge: 17.99,
            subtotal: 489.95,
            taxBeforePromotions: 0,
            taxCharged: 0,
            totalCharged: 507.94,
            totalPromotions: 10.51,
            website: 'Amazon.com',
          } as Shipment,
        ];
      });
    });

    it('should return transactions from amazon', async () => {
      const result: Array<AmazonTransaction> = [];
      for await (const transaction of getAmazonTransactions()) {
        result.push(transaction);
      }

      expect(result).toEqual([
        {
          amount: -21750,
          cleared: 'cleared',
          date: '2020-01-03',
          import_id: '6128e2d42a9b71f2680e46bd90c53fcdb8fe',
          memo: 'Some Product',
          payee_name: 'This Is A Seller',
        },
        {
          amount: 75980,
          cleared: 'cleared',
          date: '2020-01-14',
          import_id: '9b026229af8f0894ddcd8b575d28859e7357',
          memo: '2 x Some kind of coat',
          payee_name: 'A Seller',
        },
        {
          amount: 75980,
          cleared: 'cleared',
          date: '2020-01-14',
          import_id: '113728ab6700037b5a466b9b01dd0806f26c',
          memo: '2 x Some kind of coat',
          payee_name: 'A Seller',
        },
        {
          amount: -17990,
          cleared: 'cleared',
          date: '2020-02-22',
          import_id: '58b86c94e423511cc8791c294f574c51f5e6',
          memo: 'Shipping for 112-1234569-3333333',
          payee_name: 'Amazon.com',
        },
        {
          amount: 10510,
          cleared: 'cleared',
          date: '2020-02-22',
          import_id: 'ab178dd06eeec28f129b1988305e0d4ee01d',
          memo: 'Promotion for 112-1234569-3333333',
          payee_name: 'Amazon.com',
        },
      ]);
    });

    it('should respect cleared configuration', async () => {
      const config = getConfig();
      jest.spyOn(config, 'cleared', 'get').mockReturnValueOnce(false);

      const result: Array<AmazonTransaction> = [];
      for await (const transaction of getAmazonTransactions()) {
        result.push(transaction);
      }

      expect(result[0]).toHaveProperty('cleared', 'uncleared');
    });

    it('should use different import_ids for identical items', async () => {
      const result: Array<AmazonTransaction> = [];
      for await (const transaction of getAmazonTransactions()) {
        result.push(transaction);
      }

      expect(result[1].import_id).not.toEqual(result[2]);
    });

    it('should use credentials from config', async () => {
      await getAmazonTransactions().next();

      // eslint-disable-next-line
      const [[{ username, password, otpFn }]] = AmazonOrderReportsApi.mock.calls as any;
      await expect(username()).resolves.toEqual('user@example.com');
      await expect(password()).resolves.toEqual('pass123456');
      await expect(otpFn()).resolves.toEqual('54321');
    });

    it('should prompt for username/password/otp if not provided in config', async () => {
      const config = getConfig();
      jest
        .spyOn(config, 'amazonUsername', 'get')
        .mockReturnValueOnce((undefined as unknown) as string);
      jest
        .spyOn(config, 'amazonPassword', 'get')
        .mockReturnValueOnce((undefined as unknown) as string);
      jest
        .spyOn(config, 'amazonOtpCode', 'get')
        .mockReturnValueOnce((undefined as unknown) as string);

      mocked(inquirer.prompt)
        .mockResolvedValueOnce({ p: 'test@example.com' })
        .mockResolvedValueOnce({ p: 'password1234' })
        .mockResolvedValueOnce({ p: '384739' });

      await getAmazonTransactions().next();

      // eslint-disable-next-line
      const [[{ username, password, otpFn }]] = AmazonOrderReportsApi.mock.calls as any;
      await expect(username()).resolves.toEqual('test@example.com');
      await expect(password()).resolves.toEqual('password1234');
      await expect(otpFn()).resolves.toEqual('384739');
    });

    it('should use cached cookies', async () => {
      const cookies = [
        {
          name: 'session-id-time',
          value: '2082787201l',
          domain: '.amazon.com',
          path: '/',
          expires: 1641091354.410037,
          size: 26,
          httpOnly: false,
          secure: false,
          session: false,
        },
      ];

      mocked(readCache).mockResolvedValueOnce(cookies);
      await getAmazonTransactions().next();
      expect(AmazonOrderReportsApi).toBeCalledWith(
        expect.objectContaining({
          cookies,
        }),
      );
    });

    it('should save cookies', async () => {
      const cookies = [
        {
          name: 'session-id-time',
          value: '2082787201l',
          domain: '.amazon.com',
          path: '/',
          expires: 1641091354.410037,
          size: 26,
          httpOnly: false,
          secure: false,
          session: false,
        },
      ];

      await getAmazonTransactions().next();
      // eslint-disable-next-line
      const [[{ saveCookiesFn }]] = AmazonOrderReportsApi.mock.calls as any;
      await saveCookiesFn(cookies);

      expect(writeCache).toBeCalledWith('cookies', cookies);
    });
  });
});
