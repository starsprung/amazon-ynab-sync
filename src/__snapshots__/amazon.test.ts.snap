// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amazon getAmazonTransactions should handle unaccounted for adjustments 1`] = `
Array [
  Object {
    "amount": -23750,
    "cleared": "cleared",
    "date": "2020-01-04",
    "import_id": "075048d916825036a0e29dfae3ec77702be3",
    "memo": "Another Product",
    "payee_name": "This Is A Seller",
  },
  Object {
    "amount": -2500,
    "cleared": "cleared",
    "date": "2020-01-04",
    "import_id": "31e1b15b3ef2c0923e6f9bfb4db707e36d7f",
    "memo": "Shipping for 123-7654321-1234568",
    "payee_name": "Amazon.com",
  },
  Object {
    "amount": 1000,
    "cleared": "cleared",
    "date": "2020-01-04",
    "import_id": "9f796daf73a1963c7fd02bbf4dce613abae7",
    "memo": "Promotion for 123-7654321-1234568",
    "payee_name": "Amazon.com",
  },
  Object {
    "amount": -1750,
    "cleared": "cleared",
    "date": "2020-01-04",
    "import_id": "070a213531d829c843df6e25e23e8a7e6c2a",
    "memo": "Misc adjustment for 123-7654321-1234568",
    "payee_name": "Amazon.com",
  },
]
`;

exports[`amazon getAmazonTransactions should return transactions from amazon 1`] = `
Array [
  Object {
    "amount": -21750,
    "cleared": "cleared",
    "date": "2020-01-03",
    "import_id": "6128e2d42a9b71f2680e46bd90c53fcdb8fe",
    "memo": "Some Product",
    "payee_name": "This Is A Seller",
  },
  Object {
    "amount": 75980,
    "cleared": "cleared",
    "date": "2020-01-14",
    "import_id": "9b026229af8f0894ddcd8b575d28859e7357",
    "memo": "2 x Some kind of coat",
    "payee_name": "A Seller",
  },
  Object {
    "amount": 75980,
    "cleared": "cleared",
    "date": "2020-01-14",
    "import_id": "113728ab6700037b5a466b9b01dd0806f26c",
    "memo": "2 x Some kind of coat",
    "payee_name": "A Seller",
  },
  Object {
    "amount": -17990,
    "cleared": "cleared",
    "date": "2020-02-22",
    "import_id": "58b86c94e423511cc8791c294f574c51f5e6",
    "memo": "Shipping for 112-1234569-3333333",
    "payee_name": "Amazon.com",
  },
  Object {
    "amount": 10510,
    "cleared": "cleared",
    "date": "2020-02-22",
    "import_id": "ab178dd06eeec28f129b1988305e0d4ee01d",
    "memo": "Promotion for 112-1234569-3333333",
    "payee_name": "Amazon.com",
  },
]
`;
